L1Trigger/TrackFindingTMTT/interfaceDemo/DigiProducer.h

```diff
@@ -0,0 +1,69 @@
+#ifndef __DEMONSTRATOR_PRODUCER_DIGIPRODUCER_HPP__
+#define __DEMONSTRATOR_PRODUCER_DIGIPRODUCER_HPP__
+
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "DataFormats/L1TrackTrigger/interface/TTTypes.h"
+#include "DataFormats/Phase2TrackerDigi/interface/Phase2TrackerDigi.h"
+#include "SimDataFormats/TrackingAnalysis/interface/TrackingParticle.h"
+#include "SimTracker/TrackTriggerAssociation/interface/TTClusterAssociationMap.h"
+#include "SimTracker/TrackTriggerAssociation/interface/TTStubAssociationMap.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Stub.h"
+#include "L1Trigger/TrackFindingTMTT/interface/L1track3D.h"
+#include "L1Trigger/TrackFindingTMTT/interface/TrackerGeometryInfo.h"
+#include "DataFormats/Common/interface/DetSetVector.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Settings.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Histos.h"
+#include "L1Trigger/TrackFindingTMTT/interface/TrackFitGeneric.h"
+#include "Demonstrator/DataFormats/interface/DigiKF4Track.hpp"
+#include "Demonstrator/DataFormats/interface/DigiHTStub.hpp"
+#include "Demonstrator/DataFormats/interface/DigiHTMiniStub.hpp"
+#include "Demonstrator/DataFormats/interface/DigiDTCStub.hpp"
+
+#include <vector>
+#include <map>
+#include <string>
+
+using namespace std;
+
+namespace demo {
+
+  /*class Settings;
+class Histos;
+class TrackFitGeneric;*/
+
+  class DigiProducer : public edm::EDProducer {
```

Comment:
For MC production, we use python cfg that disables running Histos, since they are only needed for algorithm development. But we can't switch between using edm::stream & edm::one according to the value of this cfg param. An ugly solution would be to disable the histogram use via pragma "ifdef", which would just be enabled for development work?

Reply:
Is this producer (`demo::DigiProducer`) supposed to be run in MC production? Or is it just an example?

Reply:
L1Trigger/TrackFindingTMTT/plugins/TMTrackProducer.cc will be run during MC production . The code in interfaceDemo is used when comparing the CMSSW emulation to hardware.

Reply:
> L1Trigger/TrackFindingTMTT/plugins/TMTrackProducer.cc will be run during MC production . The code in interfaceDemo is used when comparing the CMSSW emulation to hardware.

Can the code be moved into some standard location? (e.g. `DigiProducer.h` and `DigiProducer.icc` merged into `L1Trigger/TrackFindingTMTT/plugins/DigiProducer.cc`, and `DigiConverter.h` and `DigiConverter.icc` as `DigiConverter.h` and `DigiConverter.cc` to `L1Trigger/TrackFindingTMTT/plugins` if they are not foreseen to be used outside of this package, if they are, then in `interface` and `src`)

> If you don't like the pragma solution above to allow us to disable the histograms for multithreaded use, a (more difficult) alternative would be to create EDProducts of all the classes currently passed to the histogramming class. But this would have the disadvantage of cluttering up DataFormats/ with several objects only used for optimising the L1 track performance.

I (core) would be fine with `edm::one` as a short-term solution (legacy modules should really be avoided). For a code to be run privately that could be a good-enough compromise even for longer term. But for code to be run centrally I would encourage to investigate a way to allow the use of `edm::stream` or `edm::global` base class in the long term.

Reply:
Directory interfaceDemo has been deleted